src/gridfs/download.rs

```diff
@@ -0,0 +1,181 @@
+use tokio::io::AsyncWriteExt;
+use tokio_util::compat::FuturesAsyncWriteCompatExt;
+
+use super::{options::GridFsDownloadByNameOptions, FilesCollectionDocument, GridFsBucket};
+use crate::{
+    bson::{doc, Bson},
+    error::{ErrorKind, Result},
+    options::{FindOneOptions, FindOptions},
+};
+
+impl GridFsBucket {
+    /// Downloads the contents of the stored file specified by `id` and writes
+    /// the contents to the `destination`.
+    pub async fn download_to_tokio_writer<T>(&self, id: Bson, destination: &mut T) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let options = FindOneOptions::builder()
+            .read_concern(self.read_concern().cloned())
+            .selection_criteria(self.selection_criteria().cloned())
+            .build();
+
+        let file = match self
+            .inner
+            .files
+            .find_one(doc! { "_id": &id }, options)
+            .await?
+        {
+            Some(fcd) => fcd,
+            None => {
+                return Err(ErrorKind::InvalidArgument {
+                    message: format!("couldn't find file with id {}", &id),
+                }
+                .into());
+            }
+        };
+        self.download_to_tokio_writer_common(file, destination)
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `id` and writes
+    /// the contents to the `destination`.
+    pub async fn download_to_futures_0_3_writer<T>(
+        &self,
+        id: Bson,
+        destination: &mut T,
+    ) -> Result<()>
+    where
+        T: futures_util::io::AsyncWrite + std::marker::Unpin,
+    {
+        self.download_to_tokio_writer(id, &mut destination.compat_write())
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and writes the contents to
+    /// the `destination`. If there are multiple files with the same filename, the `revision` in the
+    /// options provided is used to determine which one to download. If no `revision` is specified,
+    /// the most recent file with the given filename is chosen.
```

Comment:
Can we document in these methods that an error will be returned if the file is not found?